models/balancer/ethereum/balancer_ethereum_pools_tokens_weights.sql

```diff
@@ -0,0 +1,129 @@
+    INNER JOIN {{ source('ethereum', 'transactions') }} tx ON tx.hash = rebind.call_tx_hash 
+    WHERE rebind.call_success = TRUE
+
+    UNION ALL
+    
+    -- Unbinds
+    SELECT
+        unbind.call_block_number AS block_number, 
+        tx.index,
+        unbind.call_trace_address,
+        unbind.contract_address AS pool,
+        unbind.token,
+        '0' AS denorm
+    FROM {{ source('balancer_v1_ethereum', 'BPool_call_unbind') }} unbind
+    INNER JOIN {{ source('ethereum', 'transactions') }} tx ON tx.hash = unbind.call_tx_hash 
+    WHERE unbind.call_success = TRUE
+),
+state_with_gaps AS (
+    SELECT
+        events.block_number,
+        events.pool,
+        events.token,
+        events.denorm,
+        LEAD(events.block_number, 1) OVER (
+            PARTITION BY events.pool, events.token 
+            ORDER BY events.block_number, index, call_trace_address
+        ) AS next_block_number
+    FROM events 
+), 
+settings AS (
+    SELECT
+        pool, 
+        token, 
+        denorm
+    FROM state_with_gaps
+    WHERE
+        next_block_number IS NULL
+        AND denorm <> '0'
+),
+sum_denorm AS (
+    SELECT
+        pool,
+        SUM(denorm) AS sum_denorm
+    FROM state_with_gaps
+    WHERE
+        next_block_number IS NULL
+        AND denorm <> '0'
+    GROUP BY pool
+),
+norm_weights AS (
+    SELECT
+        settings.pool AS pool_address,
+        token AS token_address,
+        denorm / sum_denorm AS normalized_weight
+    FROM settings
+    INNER JOIN sum_denorm ON settings.pool = sum_denorm.pool
+)
+--
+-- Balancer v1 Pools Tokens Weights
+--
+SELECT
+    pool_address AS pool_id,
+    token_address,
+    normalized_weight
+FROM norm_weights
+
+UNION ALL
+
+--
+-- Balancer v2 Pools Tokens Weights
+--
+SELECT
+    registered.`poolId` AS pool_id,
+    tokens.token_address,
+    weights.normalized_weight / POWER(10, 18) AS normalized_weight
+FROM {{ source('balancer_v2_ethereum', 'Vault_evt_PoolRegistered') }} registered
+INNER JOIN {{ source('balancer_v2_ethereum', 'WeightedPoolFactory_call_create') }} call_create
+    ON call_create.call_tx_hash = registered.evt_tx_hash
+    LATERAL VIEW posexplode(call_create.tokens) tokens AS pos, token_address
+    LATERAL VIEW posexplode(call_create.weights) weights AS pos, normalized_weight
+WHERE tokens.pos = weights.pos
+
+UNION ALL
+
+SELECT
+    registered.`poolId` AS pool_id,
+    tokens.token_address,
+    weights.normalized_weight / POWER(10, 18) AS normalized_weight
+FROM {{ source('balancer_v2_ethereum', 'Vault_evt_PoolRegistered') }} registered
+INNER JOIN {{ source('balancer_v2_ethereum', 'WeightedPool2TokensFactory_call_create') }} call_create
+    ON call_create.call_tx_hash = registered.evt_tx_hash
+    LATERAL VIEW posexplode(call_create.tokens) tokens AS pos, token_address
+    LATERAL VIEW posexplode(call_create.weights) weights AS pos, normalized_weight
+WHERE tokens.pos = weights.pos
+;
```

Comment:
After moving these lines to a new file, remember to simplify the unions with a `for` loop.

Reply:
I separated the queries into v1 and v2 files.

I didn't create a `for` loop because `WeightedPoolV2Factory_call_create` has a different column name for `weights`.

Since the column names can be different it's better not to generalize these queries although they are almost the same.